### PR TITLE
Cx-Flow gets Cx-GO configuration from Cx-Integrations [328]

### PR DESCRIPTION
**Pull request by NimrodGolan**
### Description

> Allow Cx-Flow to get Cx-GO configuration (SCM access token, team & client secret) from Cx-Integrations by enabling new 'cx-integrations' configuration section.

### References

> https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/328

### Checklist

- [X] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentation is a Wiki Update, please indicate desired changes within PR MD Comment*
- [X] All active GitHub checks for tests, formatting, and security are passing
- [X] The correct base branch is being used

### CODE REVIEW

File: src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import com.checkmarx.flow.exception.MachinaRuntimeException;
+import com.checkmarx.sdk.config.Constants;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class GitAuthUrlGenerator {
+
+    public String overrideGitAuthUrlByScmAccessToken(ScanRequest.Repository repoType, String gitUrl, String scmAccessToken) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java:13]: Maybe a better name would be `addCredentialsToUrl`?

File: src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        String gitAuthUrl;
+
+        switch (repoType) {
+            case GITHUB:
+            case ADO:
+                gitAuthUrl = gitUrl.replace(Constants.HTTPS, Constants.HTTPS.concat(scmAccessToken).concat("@"));
+                return gitAuthUrl.replace(Constants.HTTP, Constants.HTTP.concat(scmAccessToken).concat("@"));
+            case BITBUCKET:
+                return gitUrl.replace(Constants.HTTPS, Constants.HTTPS.concat(scmAccessToken).concat("@"));
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java:22]: Is there an on-prem version of BitBucket? If yes, then gitUrl may start with http and not https.

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    public ReposManagerService(@Qualifier("flowRestTemplate") RestTemplate restTemplate,
+                               CxIntegrationsProperties cxIntegrationsProperties) {
+        this.restTemplate = restTemplate;
+        this.cxIntegrationsProperties = cxIntegrationsProperties;
+    }
+
+    @PostConstruct
+    private void compositeUrlPaths() {
+        cxGoConfigUrlPattern = cxIntegrationsProperties.getUrl() + "/%s/orgs/%s/tenantConfig";
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/ReposManagerService.java:35]: Why is this logic extracted into PostConstruct?

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    CxGoDynamicConfig getCxGoDynamicConfig(String scmType, String orgName) {
+        if (StringUtils.isNotEmpty(cxIntegrationsProperties.getUrl())) {
+            String urlPath = String.format(cxGoConfigUrlPattern, scmType, orgName);
+            log.info("Overriding Cx-Go configuration for SCM type: {} and organization name: {}", scmType, orgName);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/ReposManagerService.java:41]: Please make sure there is no confusion between id vs name while creating urlPath.
Also you may want to URL-encode some of the parameters.

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            ResponseEntity<CxGoDynamicConfig> responseEntity;
+            try {
+                responseEntity = restTemplate.exchange
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/ReposManagerService.java:44]: You can use restTemplate.getForEntity() here.

File: src/main/java/com/checkmarx/flow/config/external/CxGoDynamicConfig.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * Config override coming from an external web service.
+ */
+@Getter
+@Setter
+public class CxGoDynamicConfig {
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/external/CxGoDynamicConfig.java:12]: Done

File: src/main/java/com/checkmarx/flow/config/external/CxGoDynamicConfig.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    public static final String SECTION_NAME = "cxGoConfig";
+
+    @JsonProperty("team")
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/external/CxGoDynamicConfig.java:15]: Done. same fields names are now without the @JsonProperty

File: src/main/java/com/checkmarx/flow/controller/GitHubController.java [hunk 8]
```diff
@@ -342,9 +341,10 @@ public String pingRequest(
                     .excludeFiles(controllerRequest.getExcludeFiles())
                     .bugTracker(bt)
                     .filter(filter)
+                    .organizationName(StringUtils.substringBefore(repository.getFullName(), "/"))
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/controller/GitHubController.java:344]: Done.

File: src/main/java/com/checkmarx/flow/sastscanning/ScanRequestConverter.java [hunk 9]
```diff
@@ -69,31 +69,34 @@ private static boolean entityExists(Integer id) {
 
     public String determineTeamAndOwnerID(ScanRequest request) throws CheckmarxException {
         String ownerId;
-        String namespace = Optional.ofNullable(request.getNamespace()).orElse(EMPTY_STRING) ;
+        String namespace = Optional.ofNullable(request.getNamespace()).orElse(EMPTY_STRING);
 
         String team = helperService.getCxTeam(request);
         if (!ScanUtils.empty(team)) {
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/sastscanning/ScanRequestConverter.java:75]: Done.

File: src/main/java/com/checkmarx/flow/sastscanning/ScanRequestConverter.java [hunk 10]
```diff
@@ -108,6 +111,14 @@ public String determineTeamAndOwnerID(ScanRequest request) throws CheckmarxExcep
         return ownerId;
     }
 
+    private String determineOwnerId(ScanRequest request, String team) throws CheckmarxException {
+        String ownerId;
+        ownerId = (request.getClientSecret() != null)
+                ? scannerClient.getTeamIdByClientSecret(team, request.getClientSecret())
+                : scannerClient.getTeamId(team);
+        return ownerId;
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/sastscanning/ScanRequestConverter.java:119]: Done.

File: src/main/java/com/checkmarx/flow/config/CxIntegrationsProperties.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+@Component
+@ConfigurationProperties(prefix = "cx-integrations")
+@Validated
+@Getter
+@Setter
+public class CxIntegrationsProperties {
+
+    private String url;
+    private boolean readMultiTenantConfiguration;
```
[the remark this answers is not in the document]
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/config/CxIntegrationsProperties.java:17]: I left it for now as it is, as the user story defines it.

File: src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java (hunk 1 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java:13]: Done.

File: src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java (hunk 2 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/GitAuthUrlGenerator.java:22]: Right... Added also BitBucketServer case.

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java (hunk 3 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/ReposManagerService.java:35]: I found it useful, similarly to Asaf repos manager's logic

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java (hunk 5 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/ReposManagerService.java:44]: Done.

File: src/main/java/com/checkmarx/flow/service/ReposManagerService.java (hunk 4 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/ReposManagerService.java:43]: Done.

File: src/test/resources/cucumber/features/componentTests/cxIntegrations.feature [hunk 12]
```diff
@@ -0,0 +1,8 @@
+@Cx-integrations
+Feature: Cx-Flow components tests integration with Cx-integrations
+
+  Scenario: Cx-Go configuration is getting override when read-multi-tenant-configuration
+  flag in cx-integrations configuration section is on true status
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/componentTests/cxIntegrations.feature:5]: This line seems to be redundant.

File: src/test/resources/cucumber/features/componentTests/cxIntegrations.feature [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+    Given read-multi-tenant-configuration flag is on true status
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/componentTests/cxIntegrations.feature:6]:
```suggestion
    Given read-multi-tenant-configuration flag is set to true
```

File: src/test/resources/cucumber/features/componentTests/cxIntegrations.feature [hunk 14]
```diff
@@ -0,0 +1,8 @@
+@Cx-integrations
+Feature: Cx-Flow components tests integration with Cx-integrations
+
+  Scenario: Cx-Go configuration is getting override when read-multi-tenant-configuration
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/componentTests/cxIntegrations.feature:4]: Probably we should mention a web service somewhere in this scenario.

**Review by alex-ko-dev — APPROVED** [review]

**Pull request merged**